### ISSUE TITLE
question abount reward guidance and dynamics discriminator

### ISSUE THREAD

**Issue by return-sleep** [Liang-ZX/AdaptDiffuser#4]
Thanks for your excellent work. After reading the paper, I have some confusing points including treward guidance and dynamics discriminator on locomotion tasks.
1. **dynamics discriminator**
How should this dyanmics transition  T be obtained, thus allowing to get the next state based on the current state and action s_{t+1}=T(s_t,a_t). Have you performed this action and interacted with the environment online?
![20df9589c0a277ea5f84b7ecbe484d8](https://github.com/Liang-ZX/adaptdiffuser/assets/74049270/8baef310-6103-4583-b433-1570ac85aa99)
2. **reward function**
When the reward of a goal-conditioned task can be obtained by calculating the distance to the goal state. How do you  get the reward function for locomotion tasks and other unseen tasks? Does this require more information than diffuser? 
![a536a4cd82632c565efb554eff37ceb](https://github.com/Liang-ZX/adaptdiffuser/assets/74049270/bc61d5c4-07cc-4b5d-b664-2c5471c85f1b)
3 From the experimental part, is the **generalization ability** of this work only for tasks where the reward function can be known in advance (goal-conditioned task)?
In summary, I'm very curious if the generalization and adaptability of adaptdiffuser relies on extra knowledge of the task, such as transition model T and predefined reward functions, which are not available for other methods. 

Looking forward to your reply！

**Comment by Looomo** [comment]
Same question, have you found the codes for  dynamics discriminator?

**Comment by Liang-ZX** [comment]
> return-sleep

1. In different settings, we deploy different transition models. For the kuka, because the environment wrapper splits the environment interaction from the position transition of robot-arm. In kuka, we use the provided modified env.step to get next state (There are no environment interaction here). But in Mujoco locomotion, it's needed to train another transition model.

2. For locomotion, another reward model is also needed to be learned using the existing offline data. We follow the similar settings in https://github.com/jannerm/diffuser/blob/main/diffuser/models/diffusion.py#L235

3. From the replies above, it's feasible for unknown reward functions and transition model. But another seperate network is required to be trained to capture the pattern.

Thank you!

**Comment by Liang-ZX** [comment]
> Thanks for your excellent work. After reading the paper, I have some confusing points including treward guidance and dynamics discriminator on locomotion tasks.
> 
> 1. **dynamics discriminator**
>    How should this dyanmics transition  T be obtained, thus allowing to get the next state based on the current state and action s_{t+1}=T(s_t,a_t). Have you performed this action and interacted with the environment online?
>    ![20df9589c0a277ea5f84b7ecbe484d8](https://private-user-images.githubusercontent.com/74049270/287528473-8baef310-6103-4583-b433-1570ac85aa99.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5ODEyOTksIm5iZiI6MTcwMjk4MDk5OSwicGF0aCI6Ii83NDA0OTI3MC8yODc1Mjg0NzMtOGJhZWYzMTAtNjEwMy00NTgzLWI0MzMtMTU3MGFjODVhYTk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDEwMTYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmYmZhOGVlNTk4MmJlNzMyOGQyMTI2MjVmY2U1NjYwY2RjMzIxYmQ5NmZkMmMyYjY4MDRkZTM5ZDM3ZmMyOTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OJy4w6KkcY_1iFXjgSmQKyci5weXbkuEV1zQVin4hA8)
> 2. **reward function**
>    When the reward of a goal-conditioned task can be obtained by calculating the distance to the goal state. How do you  get the reward function for locomotion tasks and other unseen tasks? Does this require more information than diffuser?
>    ![a536a4cd82632c565efb554eff37ceb](https://private-user-images.githubusercontent.com/74049270/287529123-bc61d5c4-07cc-4b5d-b664-2c5471c85f1b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5ODEyOTksIm5iZiI6MTcwMjk4MDk5OSwicGF0aCI6Ii83NDA0OTI3MC8yODc1MjkxMjMtYmM2MWQ1YzQtMDdjYy00YjVkLWI2NjQtMmM1NDcxYzg1ZjFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE5VDEwMTYzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlYmI0Yjc3NWNkOTI1MzRlMjJhOTQ2Zjc2ZWRkODI2Yzk5ZTk3NGQ1MWFlYzlkNzc3ZjkxMDU4OTAxZWZkNjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4eB9wakYueXtFjjKZhqRe4vBf9uS0guatjKa1iXBo2I)
>    3 From the experimental part, is the **generalization ability** of this work only for tasks where the reward function can be known in advance (goal-conditioned task)?
>    In summary, I'm very curious if the generalization and adaptability of adaptdiffuser relies on extra knowledge of the task, such as transition model T and predefined reward functions, which are not available for other methods.
> 
> Looking forward to your reply！

1. In different settings, we deploy different transition models. For the kuka, because the environment wrapper splits the environment interaction from the position transition of robot-arm. In kuka, we use the env.step (the function provided in our code) to get next state (There are no environment interaction here).

2. For locomotion, another reward model is also required that learns from the existing offline data. We follow the similar settings in https://github.com/jannerm/diffuser/blob/main/diffuser/models/diffusion.py#L235

3. From the replies above, it's feasible for unknown reward functions and transition model. But another networks are required to capture the patterns.

Thank you!

**Comment by Looomo** [comment]
Thanks for the reply! One more question, is the transition model trained sperately with the trajectory generation model? Besides, when would the codes for locomotion tasks available?

**Comment by Liang-ZX** [comment]
> Thanks for the reply! One more question, is the transition model trained sperately with the trajectory generation model? Besides, when would the codes for locomotion tasks available?

For a simplified dynamics model, you can refer to https://github.com/anagabandi/nn_dynamics/blob/master/dynamics_model.py

And the code will be released after an updated journal version of this paper. Thanks a lot.
